xgcm/test/test_integrate.py

```diff
@@ -0,0 +1,355 @@
+            [
+                "dx_ne",
+                "dx_n",
+                "dx_e",
+                "dx_t",
+                "dy_ne",
+                "dy_n",
+                "dy_e",
+                "dy_t",
+                "dz_t",
+                "dz_w",
+                "area_ne",
+                "area_n",
+                "area_e",
+                "area_t",
+                "volume_t",
+            ],
+            [
+                dx_ne,
+                dx_n,
+                dx_e,
+                dx_t,
+                dy_ne,
+                dy_n,
+                dy_e,
+                dy_t,
+                dz_t,
+                dz_w,
+                area_ne,
+                area_n,
+                area_e,
+                area_t,
+                volume_t,
+            ],
+        ):
+            obj.coords[name] = data
+            obj.coords[name].attrs["tracked_name"] = name
+        # add xgcm attrs
+        for ii in ["xu", "xt"]:
+            obj[ii].attrs["axis"] = "X"
+        for ii in ["yu", "yt"]:
+            obj[ii].attrs["axis"] = "Y"
+        for ii in ["zt", "zw"]:
+            obj[ii].attrs["axis"] = "Z"
+        for ii in ["time"]:
+            obj[ii].attrs["axis"] = "T"
+        for ii in ["xu", "yu", "zw"]:
+            obj[ii].attrs["c_grid_axis_shift"] = 0.5
+        return obj
+
+    coords = {
+        "X": {"center": "xt", "right": "xu"},
+        "Y": {"center": "yt", "right": "yu"},
+        "Z": {"center": "zt", "right": "zw"},
+    }
+
+    metrics = {
+        ("X",): ["dx_t", "dx_n", "dx_e", "dx_ne"],
+        ("Y",): ["dy_t", "dy_n", "dy_e", "dy_ne"],
+        ("Z",): ["dz_t", "dz_w"],
+        ("X", "Y"): ["area_t", "area_n", "area_e", "area_ne"],
+        ("X", "Y", "Z"): ["volume_t"],
+    }
+    # Used previously for some fail test. Retain for now.
+    # coords_outer = {
+    #     "X": {"center": "xt", "outer": "xu"},
+    #     "Y": {"center": "yt", "outer": "yu"},
+    #     "Z": {"center": "zt", "outer": "zw"},
+    # }
+
+    # combine to different grid configurations (B and C grid)
+    ds_b = _add_metrics(
+        xr.Dataset(
+            {"u": u_b, "v": v_b, "wt": wt, "tracer": tr, "timeseries": timeseries}
+        )
+    )
+    ds_c = _add_metrics(
+        xr.Dataset(
+            {"u": u_c, "v": v_c, "wt": wt, "tracer": tr, "timeseries": timeseries}
+        )
+    )
+
+    # Some fail tests used in the weighted mean (retain for later)
+    # ds_fail = _add_metrics(
+    #     xr.Dataset(
+    #         {"u": u_b, "v": v_c, "tracer": tr, "timeseries": timeseries}
+    #     )
+    # )
+    # ds_fail2 = _add_metrics(
+    #     xr.Dataset(
+    #         {"u": u_b, "v": v_c, "tracer": tr, "timeseries": timeseries}
+    #     )
+    # )
+
+    return {
+        "B": ds_b,
+        "C": ds_c,
+        # "fail_gridtype": ds_fail,
+        # "fail_dimtype": ds_fail2,
+        "coords": coords,
+        "metrics": metrics,
+        # "fail_coords": coords_outer,
+    }
+
+
```

Comment:
I spent quit a bit of time on these tests. Could you move them into #136? That would confirm that all bugs mentioned here are 100% solved by that commit. Then I can remove all the metric code from this PR.

Reply:
The problem with these tests is that they test private methods (i.e. you use the private method `._assign_metrics` and check agains the `._metrics` attribute). I realize that other parts of xgcm do this, but I have since realized it is very bad practice. We should test the user-facing functions. The private methods are implementation details, and we should feel free to change them without breaking tests.

I will try to refactor your tests to keep the functionality without doing this.

Reply:
Ah ok I wasnt aware of that. I just wanted to ensure that everything works on the most basic level.

Reply:
Curious to see how to do it properly.